build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/CreateFrameworkList.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Xml.Linq;
+
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.Android.Tools.BootstrapTasks
+{
+	public class CreateFrameworkList : Task
+	{
+		[Required]
+		public                  ITaskItem       FrameworkListFile           { get; set; }
+
+		[Required]
+		public                  ITaskItem       FrameworkDirectory          { get; set; }
+
+		[Required]
+		public                  ITaskItem[]     FrameworkFiles              { get; set; }
+
+		public                  ITaskItem[]     FrameworkFileOverrides      { get; set; }
+
+		[Required]
+		public                  string          Redist                      { get; set; }
+
+		[Required]
+		public                  string          Name                        { get; set; }
+
+		public override bool Execute()
+		{
+			var files = new List<ITaskItem>(FrameworkFiles);
+			files.Sort ((x, y) => {
+				var a = Path.GetFileNameWithoutExtension (x.ItemSpec);
+				var b = Path.GetFileNameWithoutExtension (y.ItemSpec);
+				return string.Compare (a, b, StringComparison.OrdinalIgnoreCase);
+			});
+			var contents = new XElement ("FileList",
+					new XAttribute ("Redist", Redist),
+					new XAttribute ("Name", Name),
+					files.Select (f => ToFileElement (f)));
+			contents.Save (FrameworkListFile.ItemSpec);
+			return true;
+		}
+
+		XElement ToFileElement (ITaskItem file)
+		{
+			var path = Path.Combine (FrameworkDirectory.ItemSpec, file.ItemSpec);
+			if (!File.Exists (path)) {
+				path = Path.Combine (FrameworkDirectory.ItemSpec, "Facades", file.ItemSpec);
+			}
+			using (var pe = new PEReader (File.OpenRead (path))) {
+				var reader          = pe.GetMetadataReader();
+				var assemblyDef     = reader.GetAssemblyDefinition();
+
+				var taskVersion     = Nullable (file.GetMetadata ("Version"));
+				var overrideVersion = Nullable (FrameworkFileOverrides.FirstOrDefault (o => o.ItemSpec == file.ItemSpec)?.GetMetadata ("Version"));
+				var assemblyVersion = assemblyDef.Version.ToString ().TrimStart ('v');
+				var version         = taskVersion ?? overrideVersion ?? assemblyVersion;
+
+				var assemblyName    = new AssemblyName();
+				assemblyName.SetPublicKey (reader.GetBlobBytes (assemblyDef.PublicKey));
+
+				var publicKeyToken  = string.Join ("", assemblyName.GetPublicKeyToken ().Select(b => b.ToString ("x2")));
+
+				return new XElement ("File",
+						new XAttribute ("AsssemblyName", Path.GetFileNameWithoutExtension (file.ItemSpec)),
+						new XAttribute ("Version", version),
+						new XAttribute ("PublicKeyToken", publicKeyToken),
+						new XAttribute ("ProcessorArchitecture", "MSIL"));
+			}
+		}
+
+		static string Nullable (string value)
```

Comment:
A short method like this I normally do:
```suggestion
		static string Nullable (string value) => string.IsNullOrEmpty (value) ? null : value;
```
Up to you if you like that, though.